File: array/MonotonicQueue/cpp/MonotonicQueue.cpp
#include "MonotonicQueue.hpp"

#include <deque>

using namespace std;

vector<long long> SlidingWindowMaximum(const vector<long long>& values, int windowSize) {
    deque<int> candidateIndices;
    vector<long long> maxima;
    maxima.reserve(values.size() >= static_cast<size_t>(windowSize)
                       ? values.size() - static_cast<size_t>(windowSize) + 1
                       : 0);

    for (int i = 0; i < static_cast<int>(values.size()); ++i) {
        while (!candidateIndices.empty() &&
               candidateIndices.front() <= i - windowSize) {
            candidateIndices.pop_front();
        }

        while (!candidateIndices.empty() &&
               values[candidateIndices.back()] <= values[i]) {
            candidateIndices.pop_back();
        }

        candidateIndices.push_back(i);

        if (i + 1 >= windowSize) {
            maxima.push_back(values[candidateIndices.front()]);
        }
    }

    return maxima;
}

File: array/MonotonicQueue/java/MonotonicQueue.java
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;

public final class MonotonicQueue {
    private MonotonicQueue() {}

    public static List<Long> slidingWindowMaximum(List<Long> values, int windowSize) {
        Deque<Integer> candidateIndices = new ArrayDeque<>();
        List<Long> maxima = new ArrayList<>(Math.max(0, values.size() - windowSize + 1));

        for (int index = 0; index < values.size(); ++index) {
            while (!candidateIndices.isEmpty() && candidateIndices.peekFirst() <= index - windowSize) {
                candidateIndices.removeFirst();
            }

            while (!candidateIndices.isEmpty()
                    && values.get(candidateIndices.peekLast()) <= values.get(index)) {
                candidateIndices.removeLast();
            }

            candidateIndices.addLast(index);

            if (index + 1 >= windowSize) {
                maxima.add(values.get(candidateIndices.peekFirst()));
            }
        }

        return maxima;
    }
}
